hack/run-hive-locally.sh

```diff
@@ -0,0 +1,365 @@
+        get_envs "$type_name" "$hive_controllers_ns"
+        get_envs_from "$type_name" "$hive_controllers_ns"
+    done
+}
+
+function wait_until_hiveconfig_spec_changes {
+    local hiveconfig_spec
+    hiveconfig_spec="$(oc get hiveconfig hive -o jsonpath='{.spec}')"
+    while read -r; do
+        new_hiveconfig_spec="$(oc get hiveconfig hive -o jsonpath='{.spec}')"
+        # As "oc get -w" can give spurious events, here we make sure HiveConfig.spec is actually updated
+        if [[ "$new_hiveconfig_spec" != "$hiveconfig_spec" ]]; then
+            hiveconfig_spec="$new_hiveconfig_spec"
+            break
+        fi
+    done < <(oc get hiveconfig hive --watch-only --output-watch-events)
+    hc_spec_generation=$(( hc_spec_generation + 1 ))
+    hl_echo "Changes to HiveConfig.spec detected"
+    hl_echo "HiveConfig.spec generation = $hc_spec_generation"
+}
+
+function enable_maintenance_mode {
+    hl_echo "Enabling maintenance mode"
+    oc patch hiveconfig hive -p '{"spec": {"maintenanceMode": true}}' --type merge
+
+    hl_echo "Waiting for the controllers to be scaled down"
+    # Blocks until the existing hive-controllers is updated to have 0 replica
+    oc wait -n "$hive_controllers_ns" --for=jsonpath='{.spec.replicas}'=0 deploy/"$hive_component_controllers"
+    # Blocks until the new hive-clustersync is created and has zero replica
+    while ! oc wait -n "$hive_controllers_ns" --for=jsonpath='{.spec.replicas}'=0 sts/"$hive_component_clustersync"; do
+        sleep 1
+    done
+}
+
+function disable_maintenance_mode {
+    hl_echo "Disabling maintenance mode"
+    oc patch hiveconfig hive -p '{"spec": {"maintenanceMode": false}}' --type merge
+
+    hl_echo "Waiting for the controllers to be scaled up"
+    # Blocks until the existing hive-controllers is updated to have 1 replica
+    while ! oc wait -n "$hive_controllers_ns" --for=jsonpath='{.status.updatedReplicas}'=1 deploy/"$hive_component_controllers"; do
+        sleep 1
+    done
+    # Blocks until the new hive-clustersync is created and has the correct number of replicas
+    while ! oc wait -n "$hive_controllers_ns" --for=jsonpath='{.status.updatedReplicas}'="$hive_clustersync_replicas" sts/"$hive_component_clustersync"; do
+        sleep 1
+    done
+}
+
+function run_hive_operator_locally {
+    hl_echo "Scaling down Deployment/hive-operator"
+    oc scale -n "$hive_operator_ns" deploy/"$hive_component_operator" --replicas=0
+    oc wait -n "$hive_operator_ns" --for=jsonpath='{.spec.replicas}'=0 deploy/"$hive_component_operator"
+
+    hl_echo "Running $component locally, press ctrl + c when you are done"
+    local final_env_values=(
+        HIVE_OPERATOR_NS="$hive_operator_ns" \
+        LOG_LEVEL="$hive_operator_log_level"
+    )
+    env "${final_env_values[@]}" make run-operator 2>&1 | tee "./$artifacts_folder/${hive_component_operator}-logs.txt"
+}
+
+function run_hive_controllers_locally {
+    while true; do
```

Comment:
I don't think I understand why this loop is here. Need to look closer tomorrow...

Reply:
Sorry for the confusion. 

This loop is meant to work as follows:
- Get the current in-cluster configurations (ConfigMaps, env vars)
- Enable maintenance mode so the in-cluster controllers are scaled down
- Run the controllers locally in the background (the `make run` and `tee` things do not hang)
- `wait_until_hiveconfig_spec_changes`, which blocks until HiveConfig.spec is updated. 
This is not a hotloop here as the while loop in `wait_until_hiveconfig_spec_changes` proceeds to the next iteration only when `oc get --watch-only` gives a new message. 
- `kill_background_jobs` if we reach here, i.e. when HiveConfig.spec changes
- Disable maintenance mode, so the in-cluster controllers are scaled up, allowing us to do `oc exec` in the next iteration when getting in-cluster configurations